### ISSUE TITLE
Copy CustomIntegrator Python attributes and methods

### ISSUE THREAD

**Issue by andrrizzi** [openmm/openmm#1911]
A problem with extending `CustomIntegrator`s in Python is that any extra attribute/method is lost after copy. Overriding the `Integrator.__copy__()` method with something like this could be a solution
```python
def __copy__(self):
    """Overwrite openmm.Integrator implementation to copy class and attributes."""
    integrator_copied = super(CustomIntegrator, self).__copy__()

    # Assign correct class instead of CustomIntegrator (i.e., restore methods).
    integrator_copied.__class__ = self.__class__

    # Copy attributes. CustomIntegrator has only 1 attribute (this),
    # everything else is part of the Python implementation.
    attributes_self = {k: v for k, v in self.__dict__.items() if k != 'this'}
    integrator_copied.__dict__.update(copy.deepcopy(attributes_self))

    return integrator_copied
```
Could this be implemented as a SWIG wrapper maybe?

Having something similar also for `__getstate__()`/`__setstate__()` would solve many serialization problems (we sometimes need to send a serialized integrator over the network), although I'm not sure how to best reconcile this with the XML format currently in use.

**Comment by peastman** [comment]
Interesting idea.  This isn't even really specific to CustomIntegrator, though in practice that seems to be the only OpenMM C++ class that commonly gets subclassed in Python.  But in principle, someone could just as easily write a subclass of CustomNonbondedForce to implement some interaction, for example.

We already override the copying and pickling methods as part of the SWIG wrapper, so it would be straightforward to modify them:

https://github.com/pandegroup/openmm/blob/master/wrappers/python/src/swig_doxygen/swig_lib/python/extend.i#L362-L379

Right now we just implement pickling with the XML version, but in principle it could store other information as you suggest.

That said, this might be more complicated to implement than you think.  Copying the Python object requires also copying the C++ object and making sure the connection between the two gets hooked up correctly.  It's less general, but it might be a lot easier to just have your CustomIntegrator subclass provide its own `__copy__()` method.

**Comment by andrrizzi** [comment]
> Copying the Python object requires also copying the C++ object and making sure the connection between the two gets hooked up correctly. It's less general, but it might be a lot easier to just have your CustomIntegrator subclass provide its own `__copy__()` method.

Good point. The implementation above should take of creating the new C++ object since `super(CustomIntegrator, self).__copy__()` goes through the `XMLSerializer`. I agree that this doesn't cover all the cases, but it should work correctly in most, and users could still provide their own `__copy__` implementation.

**Comment by andrrizzi** [comment]
This being said, I can add this to the parent `CustomIntegrator` class that we use in `openmmtools` if you prefer, no problem. I just thought this could be generally useful.

**Comment by jchodera** [comment]
I think it *would* be generally useful!

**Comment by peastman** [comment]
It would definitely be useful!  We just have to make sure it works reliably and doesn't have weird edge cases.

**Comment by peastman** [comment]
This is implemented in #3263.
